### PR TITLE
SAK-45565 - Rename MySQL8 reserved words

### PR DESCRIPTION
**Pull request by mpellicer**
Tests are currently broken, this will not be as easy as it looks like.

### CODE REVIEW

**Comment by adrianfish** [comment]
Why not just rename the original, like Sam suggested? I think he's right. Just change it to SYSTEM_TASK

**Comment by adrianfish** [comment]
Also, are we actually moving to Mysql8 for 22? 5.7 is supported until 2023 and Mariadb is 5.7 and aurora compatible.

https://www.dr-chuck.com/csev-blog/2021/03/should-sakai-evolve-towards-mysql-8-0-or-maria-db-10/

It seems a bit premature to be changing db stuff for 8 when we don't know if we'll even be using it. Or am I missing some discussion around this.

**Comment by ottenhoff** [comment]
I think we’ve expressed collective desire to support MySQL 8 sooner rather
than later!

**Comment by adrianfish** [comment]
Fair enough, just asking. If we just just change that column name on the task table do we need to make this change? I think it does make sense to just change away from using a reserved word as a column name, even quoted. The task service is in 21, but it's only a conversation. Dashboards are stealthed at the minute, too.

**Comment by mpellicer** [comment]
SAK-45565 is related to another example, GROUPS is also a protected keyword.

From my point of view I think is desirable to give freedom to the developer and quote every column name by default.

https://docs.jboss.org/hibernate/orm/5.2/userguide/html_single/Hibernate_User_Guide.html#_identifier_options
https://www.chrouki.com/posts/escape-sql-reserved-keywords-jpa-hibernate/

I think we can have more issues like this in the future with MariaDB, MySQL or Oracle, setting this property will make our definitions more robust.

What do you think?

This works fine but cannot be merged because unit tests are failing, I'd like to make a decision before fixing the tests.

**Comment by adrianfish** [comment]
I just read this post by Vlad Mihalcea, and he seems to prefer just escaping when and where we need to as it gives us more control.

https://vladmihalcea.com/escape-sql-reserved-keywords-jpa-hibernate/

However, looking at the list of new reserved words in mysql8 is pretty scary:

https://dev.mysql.com/doc/mysqld-version-reference/en/keywords-8-0.html#keywords-new-in-8-0

I see a good few column names that I've used myself in there, and I'm pretty sure others will see the same.

So, I think I agree with this change. We'll have to audit all the non hibernate stuff, too.

**Comment by ottenhoff** [comment]
Wow that's a lot of reserved words

**Comment by mpellicer** [comment]
> Wow that's a lot of reserved words

Yeah, the good thing is I'm using MySQL 8 locally and I only found problems with Lessons and this new Tasks table, but I suspect Oracle can have problems too with other identifiers.

I think these new properties just make sense to support MySQL and Oracle and maybe MariaDB in the future.

https://mariadb.com/kb/en/reserved-words/

They also use SYSTEM as reserved keyword :(

**Comment by ottenhoff** [comment]
Wow this is a tough one. SYSTEM is definitely a bad column name ;) but the bigger question is how many more tables do we need to edit?

**Comment by ern** [comment]
Yeah I think we should pick a better table name, quoting is not a bad idea but prefer avoiding reserved words first. 

We would need to be aware of reprocussions of Oracle as well!
Quotes around an identifier in Oracle causes Oracle to treat the identifier as case sensitive rather than using the default of case-insensitivity. Oracle is always doing case-sensitive identifier matching, casting identifiers that are not double-quoted to upper case before doing the matching. If you put double-quotes around an identifier, Oracle skips the casting to upper case.

**Comment by mpellicer** [comment]
> Wow this is a tough one. SYSTEM is definitely a bad column name ;) but the bigger question is how many more tables do we need to edit?

As far as I can tell, only the TASKS and the LESSON_BUILDER_GROUPS tables, Sakai works just fine in MySQL 8 except these tables.

**Comment by ottenhoff** [comment]
Thanks for looking! Okay, I would vote we rename these columns for Sakai 22

**Comment by mpellicer** [comment]
This is linked with https://github.com/sakaiproject/sakai-reference/pull/120

**Comment by ottenhoff** [comment]
Awesome. Thanks @mpellicer !

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/Task.java [hunk 1]
```diff
@@ -61,7 +61,7 @@
     @Column(name = "DESCRIPTION", length = 255, nullable = false)
     private String description;
 
-    @Column(name = "\"SYSTEM\"", nullable = false)
+    @Column(name = "TASK_SYSTEM", nullable = false)
```
Comment (adrianfish) [inline kernel/api/src/main/java/org/sakaiproject/tasks/api/Task.java:64]: I think I prefer SYSTEM_TASK, tbh.

Comment (mpellicer) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/Task.java:64]: Man, SYSTEM from the TASK table is a TASK's SYSTEM right? TASK_SYSTEM

😃

Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/Task.java:64]: No, it's a task generated by the system. So, assignments creates a task, that's a system task.

Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/Task.java:64]: Really, it should have been an ENUM.

Comment (mpellicer) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/Task.java:64]: So weird man, renamed to SYSTEM_TASK.

**Pull request merged**